### PR TITLE
Openmage Designpackage and new admin theme overwrites

### PR DESCRIPTION
**Pull request by aterjung**
### Description (*)
This is a draft PR for the conversion of the openmage theme to an own package and and all related changes to the default theme to make it "infinite theme fallback" compatible. Additionally i added an own admin theme overwrite section to the backend config in Advanced/Admin/Admin Theme.
Please read issue #1216 for more details.  

### Fixed Issues
1. Fixes OpenMage/magento-lts#1216
2. Fixes OpenMage/magento-lts#1225
3. Fixes OpenMage/magento-lts#1164

### Manual testing scenarios (*)
Install branch and test it in diffenrent conditions.

### Questions or comments
1. Is this the right way to go?
2. Should the admin theme overwrites section be more failsafe?
3. Are Overwrite-Options for layout, templates,skin and locales realy needed for backend themes? There are no different storeviews with slightly different themes, you always have one backend theme.

### CODE REVIEW

File: app/code/core/Mage/Adminhtml/etc/config.xml [hunk 1]
```diff
@@ -236,23 +236,10 @@
         </help>
         <admin>
             <design>
```
Comment (kiatng) [inline app/code/core/Mage/Adminhtml/etc/config.xml:238]: `<design>` should be `<theme>` . See [system,xml](https://github.com/OpenMage/magento-lts/blob/916adc7a46c704628c61ce793ed1cde24b649be8/app/code/core/Mage/Core/etc/system.xml#L1061).

**Comment by kiatng** [comment]
The following in [Action.php](https://github.com/OpenMage/magento-lts/blob/916adc7a46c704628c61ce793ed1cde24b649be8/app/code/core/Mage/Adminhtml/Controller/Action.php#L155) breaks BC

```php
    public function preDispatch()
    {
        // override admin store design settings via admin config section
        Mage::getDesign()
            ->setArea($this->_currentArea)
            ->setPackageName((string)Mage::getConfig()->getNode('stores/admin/admin/theme/package'))
            ->setTheme((string)Mage::getConfig()->getNode('stores/admin/admin/theme/default'))
        ;
        foreach (array('layout', 'template', 'skin', 'locale') as $type) {
            if ($value = (string)Mage::getConfig()->getNode("stores/admin/admin/theme/{$type}")) {
                Mage::getDesign()->setTheme($type, $value);
            }
        }
```

Traditionally, theme overwrite is setting the package/theme in config.xml like this:

```xml
<config>
    <stores>
        <admin>
            <design>
                <package>
                    <name>default</name>
                </package>
                <theme>
                    <default>emgsori</default>
                </theme>
            </design>
        </admin>
    </stores>
    ...
</config>
```

Which will be ignored totally. Is there a way to get around this?

**Comment by aterjung** [comment]
yes, the theme overwrite is a BC in the way it is done now. This is what i think have to discussed - the way it was implementen in magento since 1.4 is a little bit strange. I work with M1 sind 2009 and never noticed that you can change the backend theme by setting the "default" value in the Design tab in Backend. I think everybody i know use this filed as default for the frontend theme.
This is the change in M1 1.4 that contains a comment showing that this was a intended decision: 
image

They decided to simply overwrite the package via config.xml. For me it seems this was forgotten right after that or it was only used for Magento Go probably - but it is not mentioned anywhere, and its not documented in the M1 Documentation.

I think the backend theme thing was not done with much love in these days - we could repair this now and add a OpenMage Theme package - then we need a way to make these BC changes work for everyone - or we keep the old code and replace the default theme.

A possible solution could be to make a switch to let the admin chose if he would like the package read from stores/admin/design/package and therefore from config.xml or from stores/admin/admin/theme/package?
On top of that - is it really needed to have the layout, theme, locale, skin inputs for the admin theme? I would prefer two dropdowns like this: 
admin_themeswitch_mockup

and if you select "legacy" everything is handled like before.

**Comment by kiatng** [comment]
For BC, we just need one dropdown for selecting legacy or openmage theme. I tested the following in [Action.php](https://github.com/OpenMage/magento-lts/blob/916adc7a46c704628c61ce793ed1cde24b649be8/app/code/core/Mage/Adminhtml/Controller/Action.php#L155):

```php
    public function preDispatch()
    {
        // User configurable setting in backend > System > Configuration > Admin > Admin Theme
        // To override openmage, create a custom package and fallback to openmage by theme.xml as in frontend.
        $switch = Mage::getStoreConfig('admin/theme/package') == 'openmage'
            ? 'admin' // Use openmage package.
            : 'design'; // Use the legacy base package for BC.

        // override admin store design settings via admin config section
        Mage::getDesign()
            ->setArea($this->_currentArea)
            ->setPackageName((string)Mage::getConfig()->getNode("stores/admin/$switch/theme/package"))
            ->setTheme((string)Mage::getConfig()->getNode("stores/admin/$switch/theme/default"))
        ;
        foreach (array('layout', 'template', 'skin', 'locale') as $type) {
            if ($value = (string)Mage::getConfig()->getNode("stores/admin/admin/theme/{$type}")) {
                Mage::getDesign()->setTheme($type, $value);
            }
        }
```

We can keep the updates on layout, template, skin, locale as is. 

If we use a `source_model` like:

```php
    public function toOptionArray()
    {
        return array(
            ['value' => 'admin', 'label'=>Mage::helper('adminhtml')->__('OpenMage theme')],
            ['value' => 'design', 'label'=>Mage::helper('adminhtml')->__('Legacy theme for backward compatibilty')],
        );
    }
```
we can simplify `$switch = Mage::getStoreConfig('admin/theme/package');`. To reduce surprises, the default should be set to legacy theme:

```xml
        <admin>
            <theme>
                <package>design</package>
            </theme>
        </admin>
```

I tested BC for `default/custom_theme`, with the override files localed in `app/design/adminhtml/default/custom_theme`, and it works.

In this way, stores with custom theme will fallback to the `base` (legacy) package. For developers, they can make custom theme to fallback to `openmage` with full flexibility as in frontend. This is the best I can think of.

**Comment by kiatng** [comment]
For existing stores that override the admin theme, and if they would like to use openmage theme and fallback to it, there is a simple way:

1. Set the `$switch` to legacy theme.
2. Add `theme.xml` to `app/design/adminhtml/default/{custom_theme}` and set the parent to `openmage/default`.

```xml
<?xml version="1.0"?>
<!--
/**
 *
 * @category    design
 * @package     default_{custom_theme}
 */
-->
<theme>
    <parent>openmage/default</parent>
</theme>
```

This works for me.

**Comment by henrykbrzoska** [comment]
Why changing to app/design/adminhtml/base ?
I'm not sure if this change 'default' to 'base' is more bad than useless.

**Comment by aterjung** [comment]
@kiatng thank you for your support on this topic! I have developed this a little further, cause i disliked that you have to switch to a "legacy mode" mode to make your own theme. The way i implemented it now is more open to the future. Building a custom theme dose not need to chose "legacy", instead you could build a custom package and it will in the dropdown to be selected.

**Comment by aterjung** [comment]
@henrykbrzoska it is a requirement to make the "infinite theme fallback" possible like it is implemented in the frontend since 1.4.
I share your doubts, but after the first experiments it seems to work flawless. But this needs more testing.

**Comment by aterjung** [comment]
@kiatng i am not so confident about the default case. 19.4.7 introduced the openmage theme as default - it would look like a regression if we change back to the default theme.

**Comment by kiatng** [comment]
@aterjung I like your approach on the switcher. It's so much better. I have tested it in following cases:

1. Test for BC as an existing store with admin theme override in package `default` theme `custom`, without `theme.xml`. It works right off the box. 
image

2. Set the config:
image

3. Added `theme.xml` in above case to fallback to openmage.
image
The fallback works as well.
image

4. As new custom package with `theme.xml` fallabck, which works perfectly as expected.




In any case, I can get BC with your latest changes. Great job!

**Comment by kiatng** [comment]
@aterjung On your concern on regressing back to the legacy by default, I feel that BC is important. As more stores upgrade/switch to om, it won't break the custom theme override. For stores without it, may be there is a way to greet them with the openmage theme?

**Comment by aterjung** [comment]
@kiatng i don't understand what you are doing there. 

```
                 $packages['default'] ?? $package = 'openmage';
```
is that intended? Or did you mean
 
```
                 $package  = $packages['default'] ?? 'openmage';
```

If not, the way you wrote it sets openmage theme if there is no default folder and the default/base package if there is a default folder present. Depending on the update strategy it is possible that a default folder is left, so this will disable the openmage theme.

I guess your code should handle the case that a shop was updated right now and there is no stores/admin/admin/theme/package entry?

Wouldn't it be better to have a setup script that runs after the update that will set this value to the best guess that is possible? 
I can imagin these cases that should be handled:

- Update from Vanilla M1
   - without custom theme -> OpenMage Theme
   - with customised theme -> Legacy Mode with Info on first Login
- Update from OpenMage 19.4.7 with OpenMage theme V1
   - with activated legacy mode -> Legacy Mode with Info on first Login
   - with activated OpenMage theme -> OpenMage Theme

Depending on the upgrade strategy we have to expect these folder structures in app/design/adminhtml/
- empty default folder
- default folder with modifications left
- default folder with complete old content from old M1 versions
- custom theme inside default folder

I think you should only set openmage theme active if we can safely identify a clean state. If not, we should activate legacy mode and show a notice about it. We should build a info page on the website where we describe what has to be done to make the openmage theme work and link it in the notice.

What do you think?

**Comment by kiatng** [comment]
@aterjung Sorry for the improper usage of null coalescing operator (??). I agree with your analysis. I think you have covered all the cases. Using setup script to set the theme is like _why-didn't-I-think-of-that_, so +1. 
+1 also on documentation and notices. With these in place, it looks perfect to me.

**Comment by aterjung** [comment]
i have now added an update script that should decide if openmage theme can be activated without danger. Only roughly tested until now...

**Comment by ADDISON74** [comment]
You did a great job please continue your work.

**Comment by fballiano** [comment]
This PR has a lot of conflicts now @aterjung are you still interested in continuing it or should we redo it from scratch?

**Comment by ADDISON74** [comment]
It can continue as a project but at least we avoid BC in OM. The list of conflicts seems to be longer and it won't be fixed soon. I express my gratitude to the efforts the author did over the years but when I see those visual issues at the top of the page I don't spend tine to fix them and switch immediately to Magento legacy. 

I propose a vote to separate this Backend theme from OM. Like what we plan with Zend Framework and CM_Redis extension.

**Comment by fballiano** [comment]
uff this PR was extremely interesting but it doesn't receive feedback since 3 years and at this point in time it's really screwed, many conflicts, the committed files would be very outdated, I think we've to close it.

I'll mark it as "don't forget this PR" and it is still linked in the issues (which will remain open). It would be easier to restart it from scratch at this point, I think.

**Comment by davirs** [comment]
I installed the latest version of 20.10.2, when I activate a theme for the frontend, the admin panel automatically switches to the legacy version. In the configuration panel, legacy is disabled.

I tested some things I saw here but without success.

image



image

**Comment by ADDISON74** [comment]
@davirs - Does your reporting refer to this closed PR or is an issue in OpenMage? If it is the last option, please post it in the Issues section, along with all the necessary steps to be reproduced and confirmed.

**Comment by davirs** [comment]
@ADDISON74 This report is related to this PR. I found other issues about exactly this problem, all closed and pointing to this PR.
Although closed, I believe that there is still some missing part of the solution.

**Comment by ADDISON74** [comment]
Please read this comment https://github.com/OpenMage/magento-lts/pull/1229#issuecomment-1546684292.  

As long as the author cannot be contacted for more than 4 years, someone else can take care of this PR if he wants to have a finality.

**Comment by sreichel** [comment]
@davirs can please open a new issue?

This PR would solve the problem, but I don't like to move that files around.

**Comment by aterjung** [comment]
Hi, sorry, i`ve completely lost this topic out of my mind and did not recoginse the messages from github. (had quite some private problems in the last years...) Is its worth getting into this topic again?

**Comment by ADDISON74** [comment]
@aterjung - This PR has been closed for more than 1 year. If you still have the strength to take it to the end, it would be better to create a new one, trying to gather all the feedback received first.

**Comment by sreichel** [comment]
> Is its worth getting into this topic again?

It is :)

**Comment by sreichel** [comment]
@aterjung please re-open, if possible.

**Pull request closed without merging**